### PR TITLE
asserts: add code for Fetcher to support the fetching of sequence-forming assertions

### PR DESCRIPTION
**Pull request by Meulengracht**
This is functionality required for https://github.com/snapcore/snapd/pull/12571 and the PRs it builds upon. In order to support fetching of sequence-forming assertions, the fetcher code needs to be extended with an additional code path.

### CODE REVIEW

File: asserts/batch.go [hunk 1]
```diff
@@ -217,7 +217,7 @@ func (b *Batch) prereqSort(db *Database) error {
 		ordered = append(ordered, a)
 		return nil
 	}
-	f := NewFetcher(db, retrieve, save)
+	f := NewFetcher(db, retrieve, nil, save)
```
Comment (mvo5) [inline asserts/batch.go:220]: I'm not super happy about this signature change as now there is a `nil` and it's hard to see what that means and also a fetcher without a retrieveSeq callback would crash (easy enough to fix of course).

So I played around a bit and was wondering if something like adding a new interface `SeqFetcher` that embedds `Fetcher` would make things a bit nicer. Full diff of my idea here: https://github.com/snapcore/snapd/compare/master...mvo5:fetcher-sequenced-alternative-idea?expand=1

So while the (unexported) fetcher implements both the Fetcher and SeqFetcher interface by returning `NewFetcher() Fetcher`  and `NewSeqFetche() SeqFetcher` interfaces we ensure that a fetcher without a retrieveSeq cannot be used to call FetchSequence.

**Review by mvo5 — COMMENTED** [review]
Thanks, this looks nice, one idea inline about how to make the signature a bit nicer (IMHO nicer, I understand this stuff is subjective :)

File: asserts/batch.go (hunk 1 above)
Comment (Meulengracht) [reply asserts/batch.go]: I really love this suggestion, I've cherry-picked your commit

**Review by mvo5 — COMMENTED** [review]
Fwiw, I like this. But I also made some substantial suggestions here so probably best to wait for Samuele to get a proper review.

File: asserts/fetcher.go [hunk 2]
```diff
@@ -60,6 +63,34 @@ func NewFetcher(trustedDB RODatabase, retrieve func(*Ref) (Assertion, error), sa
 	}
 }
 
+type SeqFetcher interface {
```
Comment (pedronis) [inline asserts/fetcher.go:66]: this probably needs a long name like SequenceFormingFetcher, also shouldn't this interface include Fetcher too?

File: asserts/fetcher.go [hunk 3]
```diff
@@ -60,6 +63,34 @@ func NewFetcher(trustedDB RODatabase, retrieve func(*Ref) (Assertion, error), sa
...
+	// FetchSequence retrieves the assertion as indicated by its sequence reference.
+	FetchSequence(*AtSequence) error
+}
+
+// NewSeqFetcher creates a Fetcher which will use trustedDB to determine trusted
```
Comment (pedronis) [inline asserts/fetcher.go:71]: this doc comment seems confusing,  trusted assertions are very likely not sequence forming. Shouldn't the first line mention SequenceFormingFetcher as that's the result type? The doc comment also doesn't talk about retrieveSeq...

File: asserts/fetcher.go [hunk 4]
```diff
@@ -60,6 +63,34 @@ func NewFetcher(trustedDB RODatabase, retrieve func(*Ref) (Assertion, error), sa
...
+// sequence forming assertions, and fetch assertions using retrieve, and then will pass
+// them to save. This is the sequence-forming variant of NewFetcher
+func NewSeqFetcher(trustedDB RODatabase, retrieve func(*Ref) (Assertion, error), retrieveSeq func(*AtSequence) (Assertion, error), save func(Assertion) error) SeqFetcher {
```
Comment (pedronis) [inline asserts/fetcher.go:74]: NewSequenceFetcher

File: asserts/fetcher.go [hunk 5]
```diff
@@ -60,6 +63,34 @@ func NewFetcher(trustedDB RODatabase, retrieve func(*Ref) (Assertion, error), sa
...
 }
 
+type SeqFetcher interface {
+	// FetchSequence retrieves the assertion as indicated by its sequence reference.
+	FetchSequence(*AtSequence) error
+}
+
+// NewSeqFetcher creates a Fetcher which will use trustedDB to determine trusted
+// sequence forming assertions, and fetch assertions using retrieve, and then will pass
+// them to save. This is the sequence-forming variant of NewFetcher
```
Comment (pedronis) [inline asserts/fetcher.go:73]: missing a dot at the end.

I'm not sure that we need the last sentence, not in this form  at least.

File: asserts/fetcher.go [hunk 6]
```diff
@@ -60,6 +63,34 @@ func NewFetcher(trustedDB RODatabase, retrieve func(*Ref) (Assertion, error), sa
...
+	return &fetcher{
+		db:          trustedDB,
+		retrieve:    retrieve,
+		retrieveSeq: retrieveSeq,
+		save:        save,
+		fetched:     make(map[string]fetchProgress),
+	}
+}
+
+func (f *fetcher) isFetched(ref *Ref) (bool, error) {
```
Comment (pedronis) [inline asserts/fetcher.go:84]: s/is/was/

File: asserts/fetcher.go [hunk 7]
```diff
@@ -106,6 +135,44 @@ func (f *fetcher) Fetch(ref *Ref) error {
 	return f.chase(ref, nil)
 }
 
+func (f *fetcher) isSeqFetched(seq *AtSequence) (bool, error) {
```
Comment (pedronis) [inline asserts/fetcher.go:138]: s/is/was/

File: asserts/fetcher.go [hunk 8]
```diff
@@ -106,6 +135,44 @@ func (f *fetcher) Fetch(ref *Ref) error {
...
+	if ok, err := f.isSeqFetched(seq); err != nil || ok {
+		// if ok is true, then the assertion was fetched and err is nil
+		return err
+	}
+	retrieved, err := f.retrieveSeq(seq)
+	if err != nil {
+		return err
+	}
+	u := seq.Unique()
+	f.fetched[u] = fetchRetrieved
```
Comment (pedronis) [inline asserts/fetcher.go:160]: sequence forming assertions could also have prerequisites so the code starting here should be shared with chase

File: asserts/fetcher.go (hunk 2 above)
Comment (Meulengracht) [reply asserts/fetcher.go]: Good points, fixed!

File: asserts/fetcher.go (hunk 3 above)
Comment (Meulengracht) [reply asserts/fetcher.go]: Updated this, thanks for the suggestion!

File: asserts/fetcher.go (hunk 4 above)
Comment (Meulengracht) [reply asserts/fetcher.go]: I took the liberty to name it `NewSequenceFormingFetcher`

File: asserts/fetcher.go (hunk 5 above)
Comment (Meulengracht) [reply asserts/fetcher.go]: I updated the entire doc string, removed the last part

File: asserts/fetcher.go [hunk 9]
```diff
@@ -60,6 +63,54 @@ func NewFetcher(trustedDB RODatabase, retrieve func(*Ref) (Assertion, error), sa
...
+	// SequenceFormingFetcher must also implement the interface of the Fetcher.
+	Fetcher
+
+	// FetchSequence retrieves the assertion as indicated by its sequence reference.
+	FetchSequence(*AtSequence) error
+}
+
+// NewSequenceFormingFetcher creates a SequenceFormingFetcher which will use trustedDB to
+// fetch sequence forming assertions using retrieveSeq, prerequisites assertions using retrieve,
+// and then will pass them to save, saving prerequisites before the depending assertions.
```
Comment (pedronis) [inline asserts/fetcher.go:76]: // NewSequenceFormingFetcher creates a SequenceFormingFetcher which will use trustedDB to determine trusted assertions,
// will fetch assertions following prerequisites using retrieve and sequence-forming assertions using retrieveSeq, and then will pass
// them to save, saving prerequisites before dependent assertions.

**Review by pedronis — COMMENTED** [review]
unit tests seems to be failing, thanks for the changes, it seems my remarks on the doc comment were confusing

File: asserts/fetcher.go (hunk 9 above)
Comment (Meulengracht) [reply asserts/fetcher.go]: Thanks for the clarifications made here, fixed.

File: asserts/fetcher_test.go [hunk 10]
```diff
@@ -164,3 +165,118 @@ func (s *fetcherSuite) TestSave(c *C) {
...
+	// Fetch same validation-set, but with a different sequence. Currently the
+	// AtSequence.Unique() does not include the sequence number or revision, meaning
+	// that the first sequence we fetch is the one that will be put into the DB.
+	// XXX: This test is here to document the behavior. If we want it to spit an error
+	//      or support multiple sequences of an assertion, then changes are required.
+	seq.Sequence = 4
+	err = f.FetchSequence(seq)
+	c.Assert(err, IsNil)
+
+	// We fetcher 2 first, it will exist
```
[the remark this answers is not in the document]
Comment (Meulengracht) [reply asserts/fetcher_test.go]: Fixed!

File: asserts/fetcher.go [hunk 11]
```diff
@@ -60,6 +63,54 @@ func NewFetcher(trustedDB RODatabase, retrieve func(*Ref) (Assertion, error), sa
 	}
 }
 
+type SequenceFormingFetcher interface {
+	// SequenceFormingFetcher must also implement the interface of the Fetcher.
+	Fetcher
+
+	// FetchSequence retrieves the assertion as indicated by its sequence reference.
```
[the remark this answers is not in the document]
Comment (Meulengracht) [reply asserts/fetcher.go]: Good point, added docs for this

**Review by MiguelPires — APPROVED** [review]
LGTM

**Review by pedronis — APPROVED** [review]
small comment and terminology issue, sequence is a either the series of assertions or the sequence value, but for the assertion at a given sequence value we use the sequence point

File: asserts/fetcher.go [hunk 12]
```diff
@@ -84,26 +136,42 @@ func (f *fetcher) chase(ref *Ref, a Assertion) error {
...
+	return f.chase(ref, nil)
+}
+
+func (f *fetcher) wasSeqFetched(seq *AtSequence) (bool, error) {
+	u := seq.Unique()
+	switch f.fetched[u] {
+	case fetchSaved:
+		return true, nil // nothing to do
+	case fetchRetrieved:
+		return false, fmt.Errorf("circular assertions are not expected: %s", seq)
```
Comment (mvo5) [inline asserts/fetcher.go:154]: I wonder if this is tested? And if not if it's worth adding a check?

File: asserts/fetcher.go [hunk 13]
```diff
@@ -63,11 +63,15 @@ func NewFetcher(trustedDB RODatabase, retrieve func(*Ref) (Assertion, error), sa
...
 }
 
+// SequenceFormingFetcher is a Fetcher with special support for fetching sequence-forming assertions through FetchSequence.
 type SequenceFormingFetcher interface {
 	// SequenceFormingFetcher must also implement the interface of the Fetcher.
 	Fetcher
 
-	// FetchSequence retrieves the assertion as indicated by its sequence reference.
+	// FetchSequence retrieves the assertion as indicated by its sequence reference. Retrieving multiple
+	// sequences of the same assertion is currently unsupported. The first sequence fetched through this
```
[the remark this answers is not in the document]
Comment (Meulengracht) [reply asserts/fetcher.go]: Good points around the terminology for sequences

File: asserts/fetcher.go (hunk 12 above)
Comment (Meulengracht) [reply asserts/fetcher.go:158]: I added tests for circular reference detection for both the original fetcher and the new! Thanks for pointing this out!

**Pull request merged**